### PR TITLE
Upgrade Babel to v7, Jest to v23.6

### PR DESCRIPTION
**Pull request by lucasterra**
This upgrades Babel to v7 and Jest to v23.6. #764 

**OTHER CHANGES**
- Fixes/updates examples which were relying on old babel configs

**BREAKING CHANGES**
- Babel v7 drops support to Node 0.10, 0.12, 4 and 5
- Jest v23.6 has some breaking changes since Jest v20.
- Updated a bunch of other dependencies as well (webpack, postcss, autoprefixer, eslint).

### CODE REVIEW

File: packages/razzle/config/env.js [hunk 1]
```diff
@@ -74,7 +74,7 @@ function getClientEnvironment(target, options) {
         PUBLIC_PATH: process.env.PUBLIC_PATH || '/',
         // CLIENT_PUBLIC_PATH is a PUBLIC_PATH for NODE_ENV === 'development' && BUILD_TARGET === 'client'
         // It's useful if you're running razzle in a non-localhost container. Ends in a /
-        CLIENT_PUBLIC_PATH: process.env.CLIENT_PUBLIC_PATH || '/',
+        CLIENT_PUBLIC_PATH: process.env.CLIENT_PUBLIC_PATH,
```
Comment (lucasterra) [inline packages/razzle/config/env.js]: This was broken since #761

File: packages/razzle/config/createConfig.js [hunk 2]
```diff
@@ -515,28 +515,16 @@ module.exports = (
...
                 // Disabled because of an issue with Uglify breaking seemingly valid code:
                 // https://github.com/facebook/create-react-app/issues/2376
                 // Pending further investigation:
                 // https://github.com/mishoo/UglifyJS2/issues/2011
                 comparisons: false,
               },
-              mangle: {
-                safari10: true,
-              },
+              mangle: true,
```
Comment (lucasterra) [inline packages/razzle/config/createConfig.js]: These were not valid uglifyjs options and was throwing errors.

File: packages/razzle/package.json [hunk 3]
```diff
@@ -29,44 +29,45 @@
...
-    "eslint-plugin-jsx-a11y": "^5.1.1",
-    "eslint-plugin-react": "^7.3.0",
-    "file-loader": "^1.1.11",
-    "fs-extra": "3.0.1",
-    "jest": "^23.1.0",
-    "mini-css-extract-plugin": "^0.4.0",
+    "@babel/core": "7.1.2",
+    "assets-webpack-plugin": "3.9.7",
+    "autoprefixer": "9.1.5",
+    "babel-core": "^7.0.0-bridge.0",
```
Comment (lucasterra) [inline packages/razzle/package.json]: This is the only addition. Everything else is version bumping.

**Comment by lucasterra** [comment]
Since this will already require a major version bump in Razzle, do you think it makes sense to remove `eslint` from the default `webpack` config and move it to a plugin instead?

Also, might be worth changing the TypeScript example/plugin to use `babel@preset-typescript` (when passing `useBabel = true`). Not only it's faster to compile, but simplifies Jest config a lot.

**Comment by olehreznichenko** [comment]
🥇

File: packages/razzle/config/env.js (hunk 1 above)
Comment (gregmartyn) [reply packages/razzle/config/env.js]: https://github.com/jaredpalmer/razzle/pull/779

**Comment by krazyjakee** [comment]
Fantastic! Would this introduce tree-shaking capabilities?

**Comment by jaredpalmer** [comment]
@lucasterra let’s remove eslint and make it a plug-in for sure

**Comment by snaerth** [comment]
Any reason this hasn´t been merged?

**Comment by krazyjakee** [comment]
@snaerth it needs to be properly reviewed by a core contributor.

@lucasterra @jaredpalmer Unless I'm mistaken, I think the eslint change could be a separate PR so that this PR could be merged and tested in the @next wild.

**Comment by lucasterra** [comment]
@jaredpalmer can you sync `next` with `master` so I can point the PR to `next` instead?

**Comment by jaredpalmer** [comment]
Can someone do this? Im on the road this week.

**Review by jaredpalmer — APPROVED** [review]

**Comment by jaredpalmer** [comment]
Just need to fix conflicts now.

**Comment by lucasterra** [comment]
Hey @jaredpalmer I think you accidentally squashed the commits from #793 they should have been merged instead.
They do have the same code, but the commit hashes are different.

Should we do a:
```
git checkout next
git reset --hard origin/master
git push -f
```

This should put `next` in sync with `master`, and won't hurt anyone, since there isn't any PRs currently pointed to `next`.

Or...
```
git checkout next
git merge origin/master
git push
```

**Comment by jaredpalmer** [comment]
done

**Comment by lucasterra** [comment]
Conflicts resolved. It's ready to merge :)

**Comment by allend13** [comment]
Merge it finally pls!

**Comment by dan-lee** [comment]
@jaredpalmer Any plans going forward with this? :)

**Comment by jaredpalmer** [comment]
Cutting a v3 release on the next branch.

**Pull request merged**